table/block_based/block_based_table_builder.cc

```diff
@@ -1182,6 +1568,22 @@ uint64_t BlockBasedTableBuilder::NumEntries() const {
 
 uint64_t BlockBasedTableBuilder::FileSize() const { return rep_->offset; }
 
+uint64_t BlockBasedTableBuilder::EstimatedFileSize() const {
+  if (rep_->compression_opts.parallel_threads > 1) {
+    std::unique_lock<std::mutex> lock(rep_->pc_rep->estimation_mutex);
```

Comment:
This function is called for each data block written. Same for the other place where the mutex is held. This would be frequent lock contention between the two threads. I'm a little bit concerned on this. Is it possible to avoid this lock while still keeping the code simply? For example, assuming `rep_->pc_rep->raw_bytes_submitted ` and `rep_->pc_rep->raw_bytes_compressed` are updated in the same thread, just protect `rep_->pc_rep->curr_compression_ratio` to be atomic?

Reply:
Replaced `raw_bytes_submit` with `raw_bytes_inflight` to get rid of mutex and minimize atomic loads in `EstimatedFileSize`. Now in `EstimatedFileSize` there are only 3 atomic variables, `rep_->offset`, `raw_bytes_inflight` and `curr_compression_ratio`.

I think we have to keep both `raw_bytes_inflight` and `curr_compression_ratio` as atomic, because these two varaibles can change in different threads.

Reply:
Maybe 3 atomic variables in `EstimatedFileSize` are still too many. I introduced a atomic variable `estimated_size` in `BlockRep`. It will be recalculated in both main thread and write thread. Now there is only 1 atomic variable in `EstimatedFileSize`.
Sorry for making a temporary change.